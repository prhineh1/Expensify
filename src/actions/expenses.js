import uuid from 'uuid';
import database from '../firebase/firebase';

//ADD_EXPENSE

export const addExpense = (expense = {}) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const addExpenseComplete = (expense = {}) => ({
    type: 'ADD_EXPENSE_COMPLETE',
    expense
});

//REMOVE_EXPENSE

export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//EDIT_EXPENSE

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//SET_EXPENSES

export const setExpenses = () => ({ type: 'SET_EXPENSES' });

export const setExpensesComplete = (expenses) => ({
    type: 'SET_EXPENSES_COMPLETE',
    expenses
});