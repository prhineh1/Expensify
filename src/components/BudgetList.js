import React from 'react';
import { connect } from 'react-redux';
import BudgetListItem from './BudgetListItem';
import moment from 'moment'; 
import partition from 'js-partition';

export class BudgetList extends React.Component {
    state = {
        displayIndex: 0
    };
    next = () => this.setState(prevstate => ({ displayIndex: prevstate.displayIndex + 1 }));
    prev = () => this.setState(prevstate => ({ displayIndex: prevstate.displayIndex - 1 }));
    render() {
        return(
            <div className='content-container'>
                <div className='list-header'>
                    <div className='show-for-mobile'>Budgets</div>
                    <div className='show-for-desktop'>Description</div>
                    <div className='show-for-desktop'>Expenses Total</div>
                    <div className='show-for-desktop'>Budgeted Amount</div>
                </div>
                <div className='list-body'>
                    {
                        this.props.budgets[this.state.displayIndex].length === 0 ? (
                            <div className='list-item list-item--message'>
                                <span>No budgets</span>
                            </div>
                        ) : (
                            this.props.budgets[this.state.displayIndex].map((budget) => 
                            <BudgetListItem 
                                key={budget.id} 
                                budget={budget} 
                                expenses={
                                    this.props.expenses.filter(expense => expense.budgetId === budget.id && 
                                    moment(expense.createdAt).isBetween(moment().startOf('month'), moment().endOf('month'), 'day'))
                                } 
                            />
                            )
                        )
                    }
                </div>
                {
                    !!this.props.budgets && (
                        <div className='list-nav'>
                            {
                                this.props.budgets[this.state.displayIndex + 1] && (
                                    <button className='button button--nav' onClick={this.next}>Next</button>
                                )
                            }
                            {  
                                this.props.budgets[this.state.displayIndex - 1] && (
                                    <button className='button button--nav' onClick={this.prev}>Previous</button>
                                )
                            }
                        </div>
                    )
                }
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    let partitionedBudgets;
    
    if (state.budgets.length % 10 === 0 && state.budgets.length > 0) {
        partitionedBudgets = partition(10, 10, state.budgets);
    }
    else {
        partitionedBudgets = partition(10, 10, [], state.budgets);        
    }
  return {
      expenses: state.expenses,
      budgets: partitionedBudgets
  };
};

export default connect(mapStateToProps)(BudgetList);