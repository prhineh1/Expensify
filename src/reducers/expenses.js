//Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE_COMPLETE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE_COMPLETE':
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE_COMPLETE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES_COMPLETE':
            return action.expenses;
        default:
            return state;
    }
};
