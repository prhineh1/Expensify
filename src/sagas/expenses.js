import ReactDOM from 'react-dom';
import { takeEvery } from 'redux-saga';
import { put, all, call } from 'redux-saga/effects'
import * as Api from '../firebase/firebase';

//WATCHERS

export function* watchAddExpenseSaga() {
    yield takeEvery('ADD_EXPENSE', addExpenseAsync);
};

export function* watchSetExpensesSaga() {
    yield takeEvery('SET_EXPENSES', setExpensesAsync);
};

export function* watchRemoveExpenseSaga() {
    yield takeEvery('REMOVE_EXPENSE', removeExpenseAsync);
};

export function* watchEditExpenseSaga() {
    yield takeEvery('EDIT_EXPENSE', editExpenseAsync);
}

//WORKERS

export function* addExpenseAsync({ expense, uid }) {
    try {
        const response = yield call(Api.create, expense, uid);  
        yield put({ 
            type: 'ADD_EXPENSE_COMPLETE', 
            expense: { id: response.key, ...expense}
        });      
    }
    catch(error) {
        return;
    }
};

export function* setExpensesAsync(uid) {
    try {
        const expenses = yield call(Api.get, uid);  
        yield put({
            type: 'SET_EXPENSES_COMPLETE',
            expenses
        });      
    }
    catch(error) {
        return;
    }
};

export function* removeExpenseAsync({ id, uid }) {
    try {
        yield call(Api.remove, id, uid);  
        yield put({
            type: 'REMOVE_EXPENSE_COMPLETE',
            id
        });      
    }
    catch(error) {
        return;
    }
};

export function* editExpenseAsync({ updates, id, uid }) {
    try {
        yield call(Api.edit, updates, id, uid); 
        yield put({
            type: 'EDIT_EXPENSE_COMPLETE',
            id,
            updates
        });       
    }
    catch(error) {
        return;
    }
};