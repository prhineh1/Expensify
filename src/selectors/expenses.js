import moment from 'moment';
import partition from 'js-partition';

//Get visible expenses

export default (expenses, {text, sortBy, startDate, endDate, pagination, expensesPartitionIndex}) => {
    let partitionedExpenses = expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return b.createdAt - a.createdAt;
        } else if (sortBy === 'amount') {
            return b.amount - a.amount;
        }
    });

    //parition(n, [step], [pad], coll)
    if (partitionedExpenses.length % pagination === 0 && partitionedExpenses.length > 0) {
        partitionedExpenses = partition(pagination, pagination, partitionedExpenses);
    }
    //when the pagination isn't a factor of partionedExpenses, add an empty array as
    //padding to ensure the remaining elements are correctly partitioned
    //if partiionedExpenses is empty, the padding ensures we get a jagged array
    else {
        partitionedExpenses = partition(pagination, pagination, [], partitionedExpenses);        
    }

    //first return value is a jagged array with all the expenses partitioned according to 'pagination'
    //second return value is the partion that will be displayed in ExpenseList.js
    return [partitionedExpenses, partitionedExpenses[expensesPartitionIndex]];
};