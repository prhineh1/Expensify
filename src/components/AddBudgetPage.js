import React from 'react';
import { connect } from 'react-redux';
import BudgetForm from './BudgetForm';
import { addBudget } from '../actions/budgets';

export class AddBudgetPage extends React.Component {
  onSubmit = (Budget) => {
    this.props.addBudget(Budget, this.props.uid);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
          <div className='page-header'>
              <div className='content-container'>
                  <h1 className='page-header__title'>Add Budget</h1>
                  <p className='page-header__subtitle'>Note: Budgeted expenses are tracked by month.</p>
              </div>
          </div>
          <div className='content-container'>
              <BudgetForm
              onSubmit={this.onSubmit}
              />
            </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({ uid: state.auth.uid });

export default connect(mapStateToProps, { addBudget })(AddBudgetPage);