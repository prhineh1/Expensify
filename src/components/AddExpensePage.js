import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense, this.props.uid)
        this.props.history.push('/');
    };
    render() {
        return (
        <div>
            <div className='page-header'>
                <div className='content-container'>
                    <h1 className='page-header__title'>Add Expense</h1>
                </div>
            </div>
            <div className='content-container'>
                <ExpenseForm
                onSubmit={this.onSubmit}
                budgets={this.props.budgets}
                />
             </div>
        </div>
        );
    }
}

const mapStateToProps = (state) => ({ 
    uid: state.auth.uid,
    budgets: state.budgets
});

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense, uid) => dispatch(addExpense(expense, uid))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);