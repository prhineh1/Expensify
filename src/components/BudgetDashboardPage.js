import React from 'react';
import BudgetsSummary from './BudgetsSummary';
import BudgetList from './BudgetList';

const BudgetDashboardPage = () => (
  <div>
      <BudgetsSummary />
      <BudgetList />
  </div>
);

export default BudgetDashboardPage;