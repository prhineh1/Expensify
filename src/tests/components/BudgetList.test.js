import React from 'react';
import { shallow } from 'enzyme';
import { BudgetList } from '../../components/BudgetList';
import expenses from '../fixtures/expenses';
import budget from '../fixtures/budgets';


test('should render BudgetList without budgets', () => {
  const wrapper = shallow(<BudgetList expenses={[]} budgets={[[]]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render BudgetList with budgets', () => {
  const wrapper = shallow(<BudgetList expenses={expenses} budgets={[[budget]]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should handle filterExpenses', () => {
  const wrapper = shallow(<BudgetList expenses={expenses} budgets={[[budget]]} />);
  const prop = wrapper.find('BudgetListItem').prop('expenses');
  expect(prop).toEqual([ expenses[1] ]);
});
