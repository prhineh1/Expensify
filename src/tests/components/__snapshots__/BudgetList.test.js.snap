// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BudgetList with budgets 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Budgets
    </div>
    <div
      className="show-for-desktop"
    >
      Description
    </div>
    <div
      className="show-for-desktop"
    >
      Expenses Total
    </div>
    <div
      className="show-for-desktop"
    >
      Budgeted Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <BudgetListItem
      budget={
        Object {
          "amount": 20000,
          "description": "utilities",
          "id": "a;sdlfj2089",
        }
      }
      expenses={
        Array [
          Object {
            "amount": 10095,
            "budgetId": "a;sdlfj2089",
            "createdAt": -345600000,
            "description": "Rent",
            "id": "2",
            "note": "",
          },
        ]
      }
      key="a;sdlfj2089"
    />
  </div>
  <div
    className="list-nav"
  />
</div>
`;

exports[`should render BudgetList without budgets 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Budgets
    </div>
    <div
      className="show-for-desktop"
    >
      Description
    </div>
    <div
      className="show-for-desktop"
    >
      Expenses Total
    </div>
    <div
      className="show-for-desktop"
    >
      Budgeted Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No budgets
      </span>
    </div>
  </div>
  <div
    className="list-nav"
  />
</div>
`;
